core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/neo/TupleInput.scala

```diff
@@ -0,0 +1,45 @@
+package edu.uci.ics.amber.engine.architecture.worker.neo
+
+import edu.uci.ics.amber.engine.common.InputExhausted
+import edu.uci.ics.amber.engine.common.tuple.ITuple
+
+class TupleInput(batchInput: BatchInput){
+
+  // save current batch related information
+  private var currentBatch:(Int, Array[ITuple]) = _
+  private var currentTupleIndex = 0
+
+  // indicate if all upstreams exhausted
+  private var allExhausted = false
+  private var inputExhaustedCount = 0
+
+  /** get next input tuple
+    * should only be called from dp thread
+    * @return tuple
+    */
+  def nextInputTuple():Either[ITuple, InputExhausted] ={
+    // increment cursor
+    currentTupleIndex += 1
+    // if batch is unavailable, take one from batchInput and reset cursor
+    if(currentBatch == null || currentTupleIndex >= currentBatch._2.length){
+      currentTupleIndex = 0
+      currentBatch = batchInput.inputBatches.take()
+    }
+    // if current batch is a data batch, return tuple
+    if(currentBatch._2 != null) {
```

Comment:
I see we are again using null to signify that it is the last message. I thought we decided to change this?